File: hackathon/security-copilot/frontend/src/types/threat.ts
export type ThreatSeverity = 'critical' | 'high' | 'medium' | 'low' | null;
export type ConfidenceLevel = 'verified' | 'high' | 'medium' | 'low';
export type ThreatStatus = 'new' | 'processing' | 'verified' | 'rejected';

export interface ThreatEntity {
  id: number;
  entity_type: 'cve' | 'vendor' | 'product' | 'vulnerability_type' | 'threat_indicator';
  value: string;
  confidence: number;
}

export interface RiskAssessment {
  risk_score: number;
  severity: string;
  cvss_score: number | null;
  exploit_available: boolean;
  public_poc: boolean;
  vendor_confirmed: boolean;
  mention_velocity: number;
  reasons: string[];
  assessed_at: string;
}

export interface ThreatEvent {
  id: number;
  external_id: string;
  source_adapter: string;
  source_type: string;
  title: string;
  description: string | null;
  url: string | null;
  author: string | null;
  published_at: string | null;
  collected_at: string;
  status: ThreatStatus;
  confidence_score: number;
  confidence_level: ConfidenceLevel;
  risk_score: number;
  severity: ThreatSeverity;
  is_duplicate: boolean;
}

export interface ThreatEventDetail extends ThreatEvent {
  confidence_reasoning: string[] | null;
  entities: ThreatEntity[];
  risk_assessment: RiskAssessment | null;
  duplicate_of_id: number | null;
  related_events: { event_id: number; similarity: number }[];
}

export interface ThreatEventListResponse {
  items: ThreatEvent[];
  total: number;
  skip: number;
  limit: number;
}

export interface ThreatStats {
  total: number;
  duplicates: number;
  by_severity: Record<string, number>;
  by_source: Record<string, number>;
  by_confidence: Record<string, number>;
}

export interface CollectorRunRequest {
  adapters?: string[] | null;
}

export interface CollectorRunResponse {
  started_at: string;
  finished_at: string | null;
  total_collected: number;
  total_inserted: number;
  total_duplicates: number;
  total_errors: number;
  adapters: Array<{
    adapter: string;
    collected: number;
    inserted: number;
    duplicates: number;
    errors: number;
    health?: Record<string, unknown>;
    error?: string;
  }>;
}

File: hackathon/security-copilot/backend/app/schemas/threat.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict


# ── Entity ────────────────────────────────────────────────────────────────────

class ThreatEntityResponse(BaseModel):
    id: int
    entity_type: str
    value: str
    confidence: float

    model_config = ConfigDict(from_attributes=True)


# ── Risk assessment ───────────────────────────────────────────────────────────

class RiskAssessmentResponse(BaseModel):
    risk_score: int
    severity: str
    cvss_score: Optional[float] = None
    exploit_available: bool
    public_poc: bool
    vendor_confirmed: bool
    mention_velocity: int
    reasons: Optional[list[str]] = None
    assessed_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Threat event list ─────────────────────────────────────────────────────────

class ThreatEventResponse(BaseModel):
    id: int
    external_id: str
    source_adapter: str
    source_type: str
    title: str
    description: Optional[str] = None
    url: Optional[str] = None
    author: Optional[str] = None
    published_at: Optional[datetime] = None
    collected_at: datetime
    status: str
    confidence_score: int
    confidence_level: str
    risk_score: int
    severity: Optional[str] = None
    is_duplicate: bool

    model_config = ConfigDict(from_attributes=True)


# ── Threat event detail ───────────────────────────────────────────────────────

class ThreatEventDetailResponse(ThreatEventResponse):
    confidence_reasoning: Optional[list[str]] = None
    entities: list[ThreatEntityResponse] = []
    risk_assessment: Optional[RiskAssessmentResponse] = None
    duplicate_of_id: Optional[int] = None
    related_events: list[dict] = []

    model_config = ConfigDict(from_attributes=True)


# ── List response ─────────────────────────────────────────────────────────────

class ThreatEventListResponse(BaseModel):
    items: list[ThreatEventResponse]
    total: int
    skip: int
    limit: int


# ── Collector ─────────────────────────────────────────────────────────────────

class CollectorRunRequest(BaseModel):
    adapters: Optional[list[str]] = None  # None = run all


class CollectorRunResponse(BaseModel):
    started_at: str
    finished_at: Optional[str] = None
    total_collected: int
    total_inserted: int
    total_duplicates: int
    total_errors: int
    adapters: list[dict]
